xdsl/transforms/stencil_to_csl_stencil.py

```diff
@@ -1,23 +1,145 @@
+
+     (1) all accesses in an arg tree are to `buf`  - arg should be moved forward in the computation
+     (2) no accesses are to `buf`                  - arg should be moved backward in the computation
+     (3) there's a mix                             - unknown, take any or no action
+
+    If two args are identified that should be moved forward, or two args are identified that should be moved backwards,
+    the computation is restructured accordingly.
+    """
+
+    buf: BlockArgument
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: arith.Addf | arith.Mulf, rewriter: PatternRewriter, /
+    ):
+        # this rewrite requires exactly 1 use which is the same type of operation
+        if len(op.result.uses) != 1 or not isinstance(
+            use := list(op.result.uses)[0].operation, type(op)
+        ):
+            return
+        c_op = use.operands[0] if use.operands[1] == op.result else use.operands[1]
+
+        def rewrite(one: Operand, two: Operand, three: Operand):
+            """
+            Builds `(one+two)+three` where `'+' == type(op)`
+            """
+            first_compute = type(op)(one, two)
+            second_compute = type(op)(first_compute, three)
+
+            # insert ops at the earliest point after both of its dependencies
+            rewriter.insert_op(
+                first_compute,
+                InsertPoint.after(
+                    max(
+                        (o.op.parent.get_operation_index(o.op), o.op)
+                        for o in [one, two]
+                        if isinstance(o, OpResult) and isinstance(o.op.parent, Block)
+                    )[1]
+                ),
+            )
+            rewriter.insert_op(
+                second_compute,
+                InsertPoint.after(
+                    max(
+                        (o.op.parent.get_operation_index(o.op), o.op)
+                        for o in [three, first_compute.results[0]]
+                        if isinstance(o, OpResult) and isinstance(o.op.parent, Block)
+                    )[1]
+                ),
+            )
+
+            rewriter.replace_op(op, [], [first_compute.results[0]])
+            rewriter.replace_op(use, [], [second_compute.results[0]])
+
+        a = get_stencil_accessed_symbols(a_op := op.lhs)
+        b = get_stencil_accessed_symbols(b_op := op.rhs)
+        c = get_stencil_accessed_symbols(c_op)
+
+        if self.move_fwd(a) and self.move_fwd(b):
+            return
+        elif self.move_back(a) and self.move_back(b):
+            return
+        elif self.move_fwd(a) and self.move_fwd(c):
+            rewrite(a_op, c_op, b_op)
+        elif self.move_fwd(b) and self.move_fwd(c):
+            rewrite(b_op, c_op, a_op)
+        elif self.move_back(a) and not self.move_back(c):
+            rewrite(c_op, b_op, a_op)
+        elif self.move_back(b) and not self.move_back(c):
+            rewrite(c_op, a_op, b_op)
```

Comment:
I _think_ this encodes some of the answers to the questions I asked in comments above. regarding ops being block args? It's a bit challenging for me to follow, if I'm honest.

Reply:
Given that :128 and :134, as well as :130 and :132 result in equivalent rewrites, I can probably shorten it at the cost of trickier-to-read conditions.